### DISCUSSION TITLE
Is this a bad practice?

### DISCUSSION THREAD

**Discussion by busslina** [GregoryConrad/rearch-dart#103] [General]
Is this a bad practice? Seems to not being working as expected, like it is not having fresh capsules on the for loop so the capsules are not being really reseted. Anywhere else on the app they are working well

```dart
DataController dataControllerCapsule(CapsuleHandle use) => DataController();

class DataController {
  Set<Capsule<llib.CachedSerializableValueController>> _cachedValueCapsules() =>
      {
        wlib.onlineWebDevicesCachedCapsule,
        voteOverallCachedCapsule,
        votePersonalCachedCapsule,
      };

  void clearData() {
    for (final capsule in _cachedValueCapsules()) {
      cc.read(capsule).reset();
    }
  }
}
```

**Comment by busslina** [comment]
A part from this issue, maybe you can recommend me a basic design in order to make all data related capsules to be self-reseted when the websocket connection state (capsule) becomes disconnected.

**Reply by busslina** [reply]
I was thinking like to use an inner listener inside each data capsule that listen to ws state and then reset themselves. But I don't know if it is possible.

**Reply by GregoryConrad** [reply]
Either:
1. Make each capsule that needs to be reset when the state changes simply listen to/use that state and invalidate itself when the state changes. In your scenario, based on what you’ve said here, this is likely preferable.
2. [Make an action capsule that groups all of the (unrelated) resets together in one reset call](https://github.com/GregoryConrad/rearch-dart/blob/241c6eb2ea10641278117b4635256a1f5b13b686/examples/scorus/lib/game_management.dart#L49)

**Reply by busslina** [reply]
It was simplier than I thought.

From this:
```dart
CachedSerializableValueController<int?> onlineWebDevicesCachedCapsule(
  CapsuleHandle use,
) {
  return use.cachedSerializableValue(null);
}
```
to this:
```dart
CachedSerializableValueController<int?> onlineWebDevicesCachedCapsule(
  CapsuleHandle use,
) {
  final ctrl = use.cachedSerializableValue(null);

  if (use(websocketStateCapsule).disconnected) {
    ctrl.reset();
  }

  return ctrl;
}
```

I remember to have already talked about being able to read other capsules from a Side Effect and you exposed me some reason on not doing that. If that were possible I would be able to read the WS state inside the side effect so all the capsules that uses it doesn't have to do that repetitive task and would be less verbose. Don't you think this case is legit to want to do that? Do you see any workaround here on the same direction?

**Reply by GregoryConrad** [reply]
To reduce the possibility of forgetting to add it and also boilerplate, I’d suggest just passing the `use(websocketStateCapsule)` into the side effect itself so it can cancel it when needed, and not have the capsule do that. Probably a `disconnected: use(websocketStateCapsule).disconnected` sort of set up.

Side effects shouldn’t really be tied to exact capsule, and keeping them capsule-less keeps them portable.

**Comment by busslina** [comment]
Never mind, the issue was related to my [last issue](https://github.com/GregoryConrad/rearch-dart/issues/97) on changing capsule states from a capsule listener . Resolved.

If you want to clarify me the secondary issues (1º about good/bad practice and 2º about listener inside a capsule) I'm open ears.

Thanks again
